src/categorical_algebra/DPO.jl

```diff
@@ -0,0 +1,215 @@
+"""
+function pushout_complement(l::ACSetTransformation,m::ACSetTransformation)::Pair{ACSetTransformation,ACSetTransformation}
+
+  I, L, G = l.dom, l.codom, m.codom
+  lm = compose(l, m)
+
+  k_components = Dict{Symbol, FinFunction}()
+  g_components = Dict{Symbol,Vector{Int}}()
+  K = ACSet{typeof(G).parameters...}()
+  offsets = Dict{Symbol,Vector{Int}}()
+
+
+  for comp in keys(l.components)
+
+    n_comp = nparts(G, comp)  # total # of elements in G
+    l_image = Set(l.components[comp].func)
+
+    # m(L/l(G)): image of (complement of image of l) into G
+    orphans = sort(map(x->m[comp](x),
+                  filter(x->!(x in l_image),
+                      1:nparts(L,comp))))
+    orph_set = Set(orphans)  # for membership test
+
+    # Compute map from G to K (undefined for orphans)
+    offsets[comp] = reindex(n_comp, orphans)
+
+    # Natural injection of subset of G into G
+    g_components[comp] = filter(x->!(x in orph_set), 1:n_comp)
+
+    # Initialize component in K
+    add_parts!(K, comp, length(g_components[comp]))
+
+    # Adjust lg function from I→G to refer to K using the offset
+    newFunc = [x - offsets[comp][x] for x in lm.components[comp].func]
+    k_components[comp] = FinFunction(newFunc, n_comp - length(orphans))
+  end
+
+  # Populate attributes for K (unchanged from G)
+  comps, arrows, src, tgt = typeof(G).parameters[1].parameters
+  attrs, srcs = typeof(G).parameters[2].parameters[3:4]
+  for (i, attr) in enumerate(attrs)
+    new=[G[attr][j] for j in g_components[comps[srcs[i]]]]
+    set_subpart!(K, attr, new)
+  end
+
+  # Populate relations for K (same as G, with offset applied)
+  for (i, col) in enumerate(arrows)
+    src_, tgt_ = comps[src[i]], comps[tgt[i]]
+    new=[val - offsets[tgt_][val] for val in G[col][g_components[src_]]]
+    set_subpart!(K, col, new)
+  end
+
+  # Put together all information into new morphisms
+  k = ACSetTransformation(I, K; k_components...)
+  @assert is_natural(k)
+  g = ACSetTransformation(K, G; g_components...)
+  @assert is_natural(g)
+
+  return k => g
+end
+
+"""
+Compute offsets induced by removing particular elements from
+the list 1:n. E.g. if we remove elements 2 and 4 from 1:5,
+our offset is [0,_,1,_,2]
+(the offset at the removed element index is not meaningful)
+"""
+function reindex(n::Int, removed::Vector{Int})::Vector{Int}
+  offset, off_counter, o_index = [],0,1
+  for i in 1:n
+    while o_index <= length(removed) && removed[o_index] < i
+      if removed[o_index] < i
+        off_counter +=1
+      end
+      o_index += 1
+    end
+    push!(offset, off_counter)
+  end
+  return offset
+end
+
+"""
+Rewrite with explicit match
+"""
+function rewrite_match(L::ACSetTransformation, R::ACSetTransformation, m::ACSetTransformation)::AbstractACSet
+    @assert L.dom == R.dom
+    @assert L.codom == m.dom
+    @assert valid_dpo(L, m)
+    @assert is_natural(L)
+    @assert is_natural(R)
+    (k, _) = pushout_complement(L, m)
+    l1, _ = pushout(R, k).cocone.legs
```

Comment:
`pushout(R, k).cocone.legs` -> `legs(pushout(R, k))` or even just `pushout(R, k)` because the colimit object is iterable.